src/key_values/key.rs

```diff
@@ -0,0 +1,143 @@
+//! Structured keys.
+
+use std::fmt;
+use std::cmp;
+use std::hash;
+use std::borrow::Borrow;
+
+/// A type that can be converted into a [`Key`](struct.Key.html).
+pub trait ToKey {
+    /// Perform the covnersion.
+    fn to_key(&self) -> Key;
+}
+
+impl<'a, T> ToKey for &'a T
+where
+    T: ToKey + ?Sized,
+{
+    fn to_key(&self) -> Key {
+        (**self).to_key()
+    }
+}
+
+impl<'k> ToKey for Key<'k> {
+    fn to_key(&self) -> Key {
+        Key {
+            key: self.key,
+        }
+    }
+}
+
+impl<'k> ToKey for &'k str {
+    fn to_key(&self) -> Key {
+        Key::from_str(self)
+    }
+}
+
+/// A key in a structured key-value pair.
+#[derive(Clone)]
+pub struct Key<'k> {
```

Comment:
So this all came out of a possible `Source::get(Key) -> Option<Value>` method, where for sources like a `BTreeMap<String, Value>` the implementation could do an indexed lookup, but for `Vec<(String, Value)>` it would need to scan. It could be a useful thing for a logging framework that wraps `log` and has some well-known structured pairs like a timestamp or correlation id that consumers want to poke at.

How you would get a hold of an index, and how useful that index would be if you, say, concat two `Source`s together is where I stopped the design train and left it as a possible extension. For the well-known pairs use-case you could also just write a `Source` like:

```rust
struct MyFrameworkSource<'s> {
    timestamp: Timestamp,
    correlation: Uuid,
    the_rest: Option<&'s dyn Source>,
}
```

All that's to say it's something I didn't want to absolutely rule out as an enhancement but it's still airy-fairy.